File: Partition array to K subsets - GFG/partition-array-to-k-subsets.java
// { Driver Code Starts
import java.util.*;

class Partition_Arr_To_K_Subsets
{
	public static void main(String args[])
	{
		Scanner sc = new Scanner(System.in);
		int t = sc.nextInt();
		while(t>0)
		{
			int n = sc.nextInt();
			int a[] = new int[n];
			for(int i=0;i<n;i++)
				a[i] = sc.nextInt();
			int k = sc.nextInt();
			Solution g = new Solution();
			if(g.isKPartitionPossible(a,n,k)==true)
				System.out.println(1);
			else 
				System.out.println(0);
			
		t--;
		}
	}
}// } Driver Code Ends


/*You are required to complete this method */
class Solution {
    public boolean isKPartitionPossible(int[] arr,int n, int k) {
        int sum=0;
        for(int i:arr) sum+=i;
        if(sum%k!=0) return false;
        Arrays.sort(arr);
        
        return (dfs(n-1,arr,new int[k],k,sum/k));
    }
    public boolean dfs(int ind,int[] arr,int[] sum, int k,int target){
        if(ind == -1)return true;
        for(int i=0;i<k;i++){
            if((sum[i] + arr[ind]>target) || (i>0 && sum[i] == sum[i-1]))continue;
            
            sum[i]+=arr[ind];
            
            if(dfs(ind-1,arr,sum,k,target))return true;
            sum[i]-= arr[ind];
        }
        return false;
    }
}

File: Partition array to K subsets - GFG/partition-array-to-k-subsets.py
#User function Template for python3

from functools import lru_cache
class Solution:
    def isKPartitionPossible(self, nums, k):
        #code here
        if sum(nums)%k:
            return False
        visited=[False]*len(nums)
        nums.sort(reverse=True)
        
        target=sum(nums)/k
        
        @lru_cache(None)
        def solve(i,k,ans):
            if k==0:
                return True
            if ans==target:
                return solve(0,k-1,0)
            for j in range(i,len(nums)):
                if visited[j] or ans+nums[j]>target:
                    continue
                visited[j]=True
                if solve(j+1,k,ans+nums[j]):
                    return True
                visited[j]=False
            return False
        return solve(0,k,0)
        
#{ 
#  Driver Code Starts


if __name__ == '__main__':
    tcs = int(input())
    for _ in range(tcs):
        N=int(input())
        arr=[int(x) for x in input().split()]
        k=int(input())
        if Solution().isKPartitionPossible(arr, k):
            print(1)
        else:
            print(0)
# } Driver Code Ends